File: ai_engine/detectors/url_detector.py
"""
URL Safety & Phishing Detector for FraudLens.
Performs lexical analysis, bank typosquatting detection, high-risk TLD checks,
and phishing path heuristic classification.
"""

import re
from urllib.parse import urlparse
from typing import Dict, Any

LEGITIMATE_BANK_DOMAINS = {
    'sbi': ['sbi.co.in', 'onlinesbi.sbi', 'onlinesbi.com'],
    'hdfc': ['hdfcbank.com', 'hdfc.com'],
    'icici': ['icicibank.com', 'icicidirect.com'],
    'axis': ['axisbank.com'],
    'pnb': ['pnbindia.in'],
    'kotak': ['kotak.com'],
    'paytm': ['paytm.com'],
    'phonepe': ['phonepe.com'],
    'googlepay': ['pay.google.com', 'google.com']
}

HIGH_RISK_TLDS = {
    '.vip', '.top', '.xyz', '.cc', '.work', '.site', '.club', '.live',
    '.tk', '.ml', '.ga', '.cf', '.gq', '.info', '.buzz', '.icu', '.fit'
}

PHISHING_KEYWORDS = [
    'kyc', 'pan', 'aadhaar', 'verify', 'update', 'login', 'secure',
    'support', 'unblock', 'reward', 'refund', 'cashback', 'lottery',
    'claim', 'apk', 'bonus'
]

class URLDetector:
    def __init__(self):
        pass

    def analyze(self, raw_url: str) -> Dict[str, Any]:
        """Analyze URL for brand typosquatting, suspicious TLDs, and credential harvesting paths."""
        url = raw_url.strip()
        if not re.match(r'^[a-zA-Z]+://', url):
            url = 'http://' + url

        try:
            parsed = urlparse(url)
            hostname = (parsed.hostname or '').lower()
            path = (parsed.path or '').lower()
            scheme = parsed.scheme.lower()
        except Exception:
            hostname = url.lower()
            path = ''
            scheme = 'http'

        flags = []
        tactics = []
        impersonated_brand = None
        risk_score = 10

        is_ip = bool(re.match(r'^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$', hostname))
        if is_ip:
            flags.append({
                "title": "Raw IP Address Hostname",
                "detail": f"The URL uses a raw numeric IP address ({hostname}) instead of an institutional domain name. Standard banks and financial providers never host consumer portals on bare IPs.",
                "severity": "critical"
            })
            tactics.append("Direct IP Obfuscation")
            risk_score += 45

        # Extract TLD
        tld = '.' + hostname.split('.')[-1] if '.' in hostname else ''
        if tld in HIGH_RISK_TLDS:
            flags.append({
                "title": f"High-Risk Top-Level Domain ({tld})",
                "detail": f"The {tld} extension has a statistically heavy correlation with disposable phishing infrastructure and botnet campaigns.",
                "severity": "high"
            })
            tactics.append("Malicious Disposable TLD")
            risk_score += 35

        # Brand Impersonation & Typosquatting Check
        for brand, legit_domains in LEGITIMATE_BANK_DOMAINS.items():
            if brand in hostname:
                is_legit = any(hostname == ld or hostname.endswith('.' + ld) for ld in legit_domains)
                if not is_legit:
                    impersonated_brand = brand.upper()
                    flags.append({
                        "title": f"Brand Impersonation & Typosquatting ({impersonated_brand})",
                        "detail": f"The domain mimics '{impersonated_brand}' but does not belong to authorized domains ({', '.join(legit_domains)}). Scammers spoof financial brands to harvest login credentials and OTPs.",
                        "severity": "critical"
                    })
                    tactics.append("Typosquatting Mimicry")
                    risk_score += 50
                    break

        # Phishing Keywords in Path or Subdomain
        matched_keywords = [kw for kw in PHISHING_KEYWORDS if kw in hostname or kw in path]
        if matched_keywords:
            flags.append({
                "title": f"Credential Harvesting Lures ({', '.join(matched_keywords[:3])})",
                "detail": f"URL contains keywords frequently weaponized in credential harvesting campaigns: {', '.join(matched_keywords)}.",
                "severity": "medium"
            })
            tactics.append("Credential Harvesting Lures")
            risk_score += 20

        # Subdomain count check
        subdomain_parts = hostname.split('.')
        if len(subdomain_parts) > 3 and not is_ip:
            flags.append({
                "title": "Excessive Subdomain Nesting",
                "detail": f"Excessive domain nesting ({len(subdomain_parts)} levels) is a common obfuscation technique to disguise fraudulent host destinations.",
                "severity": "medium"
            })
            tactics.append("Subdomain Obfuscation")
            risk_score += 15

        # Insecure protocol for financial operations
        if scheme == 'http' and (impersonated_brand or matched_keywords):
            flags.append({
                "title": "Unencrypted HTTP Communication",
                "detail": "Target link operates over unencrypted HTTP. Legitimate financial portals mandate TLS 1.3 / HTTPS encryption.",
                "severity": "medium"
            })
            risk_score += 15

        # Final score calculation
        if any(f['severity'] == 'critical' for f in flags):
            risk_score = max(risk_score, 85)

        risk_score = max(5, min(risk_score, 99))

        if risk_score >= 70:
            classification = "High Risk"
            confidence = 0.98
        elif risk_score >= 35:
            classification = "Suspicious"
            confidence = 0.85
        else:
            classification = "Safe"
            confidence = 0.95
            if not flags:
                flags.append({
                    "title": "Legitimate Institutional Domain",
                    "detail": "URL structure conforms to verified public registry standards with no typosquatting or deception signatures detected.",
                    "severity": "safe"
                })
                tactics.append("Verified Domain Integrity")

        # Explainable AI report
        summary = (
            f"FraudLens AI evaluated this link with Risk Score {risk_score}/100 ({classification}). "
            + (f"Impersonating {impersonated_brand} on an unauthorized server." if impersonated_brand else "Domain integrity checked.")
        )
        immediate_action = (
            "DO NOT open this link or enter netbanking passwords, UPI PINs, or card CVVs. Report link to CERT-In (incident@cert-in.org.in)."
            if classification != "Safe" else "Link appears safe to visit. Verify SSL lock in your browser address bar."
        )
        key_takeaway = "Banks never send shortened or disposable .vip/.top URLs for sensitive KYC verifications."

        complaint_draft = (
            f"=== NATIONAL CYBER CRIME REPORTING PORTAL (1930) DRAFT ===\n"
            f"Incident Category: Phishing Website / Banking Typosquatting\n"
            f"Suspect URL: {raw_url}\n"
            f"Impersonated Entity: {impersonated_brand or 'Unknown Entity'}\n"
            f"Threat Assessment: {classification} (Risk Score: {risk_score}/100)"
        )

        return {
            "type": "url",
            "input": raw_url,
            "riskScore": risk_score,
            "classification": classification,
            "confidence": confidence,
            "flags": flags,
            "tactics": tactics,
            "impersonatedBrand": impersonated_brand,
            "domainDetails": {
                "domain": hostname,
                "isIpAddress": is_ip,
                "tld": tld,
                "path": path,
                "protocol": scheme
            },
            "aiExplanation": {
                "summary": summary,
                "socialEngineeringTactics": tactics,
                "immediateAction": immediate_action,
                "keyTakeaway": key_takeaway,
                "complaintDraft": complaint_draft,
                "generatedByLLM": False
            }
        }

url_detector = URLDetector()

File: backend/services/urlDetector.js
/**
 * FraudLens: URL & Phishing Link Inspection Engine
 * Detects typosquatting, deceptive subdomains, suspicious TLDs, and IP address hosts.
 */

const TRUSTED_DOMAINS = [
  'onlinesbi.sbi',
  'sbi.co.in',
  'hdfcbank.com',
  'icicibank.com',
  'axisbank.com',
  'pnbindia.in',
  'bankofbaroda.in',
  'paytm.com',
  'phonepe.com',
  'google.com',
  'bhimupi.org.in',
  'npci.org.in',
  'rbi.org.in',
  'incometax.gov.in',
  'cybercrime.gov.in',
  'bescom.karnataka.gov.in'
];

const TARGETED_BRANDS = [
  'sbi', 'onlinesbi', 'yono', 'hdfc', 'icici', 'axis', 'pnb', 'bob',
  'paytm', 'phonepe', 'gpay', 'bhim', 'npci', 'aadhaar', 'uidai',
  'incometax', 'electricity', 'bescom', 'tneb', 'mseb', 'tatapower'
];

const SUSPICIOUS_TLDS = new Set([
  'vip', 'top', 'xyz', 'live', 'work', 'click', 'buzz', 'cfd',
  'link', 'gq', 'ml', 'cf', 'tk', 'icu', 'rest', 'monster', 'site', 'fun'
]);

function levenshteinDistance(s1, s2) {
  const m = s1.length;
  const n = s2.length;
  const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      } else {
        dp[i][j] = 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
      }
    }
  }
  return dp[m][n];
}

function calculateEntropy(str) {
  const map = {};
  for (let char of str) {
    map[char] = (map[char] || 0) + 1;
  }
  let entropy = 0;
  for (let char in map) {
    const p = map[char] / str.length;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

export function analyzeUrlSafety(inputUrl) {
  if (!inputUrl || typeof inputUrl !== 'string') {
    return {
      riskScore: 0,
      classification: 'Safe',
      reasons: ['No URL provided']
    };
  }

  let normalized = inputUrl.trim();
  if (!/^https?:\/\//i.test(normalized)) {
    normalized = 'http://' + normalized;
  }

  let parsed;
  try {
    parsed = new URL(normalized);
  } catch (err) {
    return {
      riskScore: 75,
      classification: 'Suspicious',
      reasons: [{ title: 'Malformed URL', detail: 'URL syntax is invalid or obfuscated.', severity: 'high' }]
    };
  }

  const hostname = parsed.hostname.toLowerCase();
  const pathname = parsed.pathname.toLowerCase();
  const flags = [];
  let score = 5;

  // 1. Check if it's already an exact trusted institutional domain
  const isDirectTrusted = TRUSTED_DOMAINS.some(td => hostname === td || hostname.endsWith('.' + td));
  if (isDirectTrusted) {
    return {
      url: normalized,
      hostname,
      riskScore: 5,
      classification: 'Safe',
      confidence: 0.98,
      isTrustedDomain: true,
      flags: [{
        title: 'Verified Official Domain',
        detail: `Matches authentic registered domain of ${hostname}.`,
        severity: 'safe'
      }],
      securityChecklist: {
        sslTls: parsed.protocol === 'https:',
        typosquattingSafe: true,
        reputableTLD: true,
        ipHostSafe: true
      }
    };
  }

  // 2. IP Address in Hostname
  const isIpAddress = /^(\d{1,3}\.){3}\d{1,3}$/.test(hostname);
  if (isIpAddress) {
    flags.push({
      title: 'Raw IP Address Host',
      detail: `URL uses a raw IP address (${hostname}) instead of a registered domain name. Major red flag for temporary phishing command servers.`,
      severity: 'critical'
    });
    score += 50;
  }

  // 3. Suspicious / Abused TLD
  const domainParts = hostname.split('.');
  const tld = domainParts[domainParts.length - 1];
  if (SUSPICIOUS_TLDS.has(tld)) {
    flags.push({
      title: `High-Risk Top-Level Domain (.${tld})`,
      detail: `The .${tld} extension has a high statistical correlation with disposable cyber fraud infrastructure according to global threat intelligence.`,
      severity: 'high'
    });
    score += 35;
  }

  // 4. Typosquatting / Brand Name Impersonation in Hostname or Path
  let brandSpoofed = null;
  TARGETED_BRANDS.forEach(brand => {
    if (hostname.includes(brand)) {
      brandSpoofed = brand.toUpperCase();
    }
  });

  if (brandSpoofed && !isDirectTrusted) {
    flags.push({
      title: `Brand Impersonation (${brandSpoofed})`,
      detail: `The domain contains "${brandSpoofed}", but is not an official domain of this institution. Scammers create spoofed domains to harvest login credentials and OTPs.`,
      severity: 'critical'
    });
    score += 55;
  }

  // 5. Check Levenshtein Distance against genuine bank domains
  TRUSTED_DOMAINS.forEach(genuine => {
    const mainGenuine = genuine.split('.')[0];
    const mainInput = domainParts[domainParts.length - 2] || domainParts[0];
    const dist = levenshteinDistance(mainGenuine, mainInput);
    if (dist >= 1 && dist <= 2 && mainInput.length > 3) {
      flags.push({
        title: `Lookalike Typosquatting of ${genuine}`,
        detail: `The domain "${hostname}" is deceptively similar to legitimate portal "${genuine}" (Levenshtein distance: ${dist}).`,
        severity: 'critical'
      });
      score += 50;
    }
  });

  // 6. Dangerous keywords in URL (KYC, verify, update-pan, download)
  const phishingKeywords = ['kyc', 'update-pan', 'verify', 'yono', 'netbanking', 'cashback', 'refund', 'login', 'free-gift'];
  const matchedKeywords = phishingKeywords.filter(kw => hostname.includes(kw) || pathname.includes(kw));
  if (matchedKeywords.length > 0) {
    flags.push({
      title: `Phishing Lure Terminology (${matchedKeywords.join(', ')})`,
      detail: `URL path contains terms commonly weaponized in credential harvesting campaigns.`,
      severity: 'medium'
    });
    score += 20 * matchedKeywords.length;
  }

  // 7. APK or Executable download (path OR hostname — scammers host raw .apk in the domain)
  if (pathname.endsWith('.apk') || pathname.endsWith('.exe') || /\.apk$|\.exe$/i.test(hostname)) {
    flags.push({
      title: 'Direct Android APK Payload Download',
      detail: 'Attempts to download an untrusted mobile application package bypassing the official Google Play Store.',
      severity: 'critical'
    });
    score += 45;
  }

  // 8. Excessive Subdomains & Entropy
  if (domainParts.length > 3) {
    flags.push({
      title: 'Excessive Subdomain Layering',
      detail: `Domain uses ${domainParts.length} levels of subdomains to visually mask the actual root domain on mobile browsers.`,
      severity: 'medium'
    });
    score += 20;
  }

  const entropy = calculateEntropy(hostname);
  if (entropy > 3.8 && hostname.length > 15) {
    flags.push({
      title: 'High Domain Name Entropy',
      detail: 'Hostname exhibits high character randomness typical of algorithmically generated domain names (DGA).',
      severity: 'medium'
    });
    score += 20;
  }

  // Clamp final score
  score = Math.min(Math.max(score, 10), 99);

  let classification = 'Safe';
  if (score >= 70) {
    classification = 'High Risk';
  } else if (score >= 35) {
    classification = 'Suspicious';
  }

  return {
    url: normalized,
    hostname,
    protocol: parsed.protocol,
    tld: `.${tld}`,
    riskScore: score,
    classification,
    confidence: +(0.85 + (score > 60 ? 0.12 : 0.05)).toFixed(2),
    flags,
    securityChecklist: {
      sslTls: parsed.protocol === 'https:',
      typosquattingSafe: !brandSpoofed,
      reputableTLD: !SUSPICIOUS_TLDS.has(tld),
      ipHostSafe: !isIpAddress
    }
  };
}
